modules/xcm-bridge-hub/src/lib.rs

```diff
@@ -0,0 +1,1883 @@
+
+				// write something to log
+				let enqueued_messages = outbound_lane.queued_messages().checked_len().unwrap_or(0);
+				log::trace!(
+					target: LOG_TARGET,
+					"Bridge {:?} between {:?} and {:?} is closing. {} messages remaining",
+					locations.lane_id,
+					locations.bridge_origin_universal_location,
+					locations.bridge_destination_universal_location,
+					enqueued_messages,
+				);
+
+				// deposit the `ClosingBridge` event
+				Self::deposit_event(Event::<T, I>::ClosingBridge {
+					lane_id: locations.lane_id,
+					pruned_messages,
+					enqueued_messages,
+				});
+
+				return Ok(())
+			}
+
+			// else we have pruned all messages, so lanes and the bridge itself may gone
+			inbound_lane.purge();
+			outbound_lane.purge();
+			Bridges::<T, I>::remove(locations.lane_id);
+			BridgesByLocalOrigin::<T, I>::try_mutate(
+				VersionedInteriorMultiLocation::from(locations.bridge_origin_universal_location),
+				|storage_bridges| {
+					let mut bridges = storage_bridges.to_vec();
+					bridges.retain(|b| *b != locations.lane_id);
+					*storage_bridges = bridges
+						.try_into()
+						.map_err(|_| Error::<T, I>::TooManyBridgesForLocalOrigin)?;
+					Ok::<(), Error<T, I>>(())
+				},
+			)?;
+
+			// unreserve remaining amount
+			let failed_to_unreserve =
+				T::NativeCurrency::unreserve(&bridge.bridge_owner_account, bridge.reserve);
+			if !failed_to_unreserve.is_zero() {
+				// we can't do anything here - looks like funds have been (partially) unreserved
+				// before by someone else. Let's not fail, though - it'll be worse for the caller
+				log::trace!(
+					target: LOG_TARGET,
+					"Failed to unreserve {:?} during ridge {:?} closure",
+					failed_to_unreserve,
+					locations.lane_id,
+				);
+			}
+
+			// write something to log
+			log::trace!(
+				target: LOG_TARGET,
+				"Bridge {:?} between {:?} and {:?} has been closed",
+				locations.lane_id,
+				locations.bridge_origin_universal_location,
+				locations.bridge_destination_universal_location,
+			);
+
+			// deposit the `BridgePruned` event
+			Self::deposit_event(Event::<T, I>::BridgePruned {
+				lane_id: locations.lane_id,
+				pruned_messages,
+			});
+
+			Ok(())
+		}
+
+		/// Report bridge misbehavior.
+		///
+		/// The origin must be signed. If misbehavior is confirmed, some (`Penalty`) portion of
+		/// bridge reserve is transferred to the reporter. The inbound XCM channel between local
+		/// origin chain and the bridge hub is suspended and the outbound lane of the bridge is
+		/// immediately closed.
+		///
+		/// The bridge owner must fix the misbehavior by processing the corresponding bridge queue.
+		/// After that, he may call (using some associated account) the `resume_misbehaving_bridges`
+		/// to return to regular bridge operations.
+		#[pallet::call_index(2)]
+		#[pallet::weight(Weight::zero())] // TODO: https://github.com/paritytech/parity-bridges-common/issues/1760 - weights
+		pub fn report_misbehavior(
```

Comment:
Two things to add:
1) if #2261 will be implemented, we'll probably just remove this function or at least the current misbehavior. That's because it will be a sending chain problem to deal with its own queue/storage. Let's keep it for now, though, until we have a confidence that #2261 is the right option;
2) there's currently a problem - if bridge is closed at one side and misbehaving at other, the messages and the bridge remains will live here forever. I'll deal with that in a follow up PRs - that is noted [here](https://github.com/paritytech/parity-bridges-common/issues/1760#issuecomment-1620163755) under the " start closing lane at this side if the same lane is in closing state at the bridged side" point

Reply:
If we are not sure about the misbehavior part yet, could we keep in this PR only the open-close bridge part ? And move the misbehavior to #2261 ? Or to a separate PR anyway ?

To me it seems easier to review and merge this way.

Reply:
Right! It doesn't belong actually to #2261, because it isn't about dynamic fees, it just depends on the way our low-level transport works. So separate PR (on top of this one) looks good - I'll extract those two calls (`report_misbehavior` and `resume_misbehaving_bridges`) soon. Thanks for suggestion

Reply:
done